### ISSUE TITLE
VirtualizedScrollPane estimatedScrollYProperty is unstable

### ISSUE THREAD

**Issue by swpalmer** [TomasMikula/RichTextFX#307]
Attempting to synchronize the scroll position of two CodeAreas is not possible because the estimatedScrollYPosition is unstable.  During a scroll operation the reported value flickers to unreasonable values.

For example, if I print estimatedScrollYPosition.value() in a listener while I drag the scrollbar down, I see values like this:

Initial values reported from listener before I grab the scroll bar:
-4.375  (this seems odd)
0.0  (this is expected)

Then as I drag:
5.6
11.2
16.8
... all good for little bit... then...
117.6
174.8 (WRONG!)
117.6  (interesting, back to the SAME value as before, not greater than 117.6)
...
130.1
244.6  (WRONG! )
130.1
...
173.6
230.9 (WRONG!)
173.6
...

So when binding the value to another scroll pane we see those mistakes visually.  Interestingly, the original pane that is being scrolled does not flicker, unless the binding is bidirectional.

**Comment by TomasMikula** [comment]
Thanks for testing this!

Just an idea: Does the appearance of the "WRONG" values correlate with a line longer than any previously seen lines entering the viewport?

**Comment by TomasMikula** [comment]
Could you also please intersperse the output with changes to `totalLengthEstimateProperty()`, to see whether the glitches in position estimate are caused by glitches in the estimated total length of the content?

**Comment by swpalmer** [comment]
The longest line is in the viewport initially, so only shorter lines are scrolling in.
The glitches happen on line boundaries.  So either a new line coming in or an old line going out is happening at the same time.

Sometimes the values are very strange.  I observed this sequence for example, while slow scrolling up and the first line started to come into view:
56.0
-1.272727
0.0
56.0

Do you mean totalHeightEstimateProperty()?  It is reported as the same value all the time.
I don't see a totalLengthEstimateProperty().

I just noticed that if I click the down arrow on the scroll bar instead of dragging, that it reproduces the glitch every time:

One click produces:
61.091
118.364
61.091

next click I get:
122.182
179.455
122.182

then:
183.273
240.545
183.273

I'm also seeing faint horizontal lines between the lines of text (rounding error?) at certain scroll positions, but that's another more minor issue.

**Comment by swpalmer** [comment]
I should note that when I observe the glitches I seem to get all three (of four in the case of 56.0, -1.273, 0.0,56.0) events in response to the scrollbar being moved a single pixel.

**Comment by TomasMikula** [comment]
So I found and fixed one bug in Flowless. You can try with the newest 1.0.0-SNAPSHOT release (or complile updated sources), which uses Flowless 0.6-SNAPSHOT.

The glitches in `estimatedScrollYProperty()` should not cause visual glitches as long as it reaches the correct value before the next layout pass, which it seems it does.

**Comment by Crystark** [comment]
FYI, tried 1.0.0-SNAPSHOT and i'm getting the following exception when i try to open the editor in my app:

```
Caused by: java.lang.NoSuchMethodError: org.fxmisc.flowless.VirtualFlow.breadthOffsetProperty()Ljavafx/beans/property/ReadOnlyDoubleProperty;
	at org.fxmisc.richtext.StyledTextArea.createCell(StyledTextArea.java:1064)
	at org.fxmisc.richtext.StyledTextArea.lambda$new$3(StyledTextArea.java:549)
	at org.fxmisc.flowless.CellPool.getCell(CellPool.java:20)
	at org.fxmisc.flowless.CellListManager.cellForItem(CellListManager.java:75)
	at org.reactfx.collection.MappedList.get(MappedList.java:27)
	at org.reactfx.collection.MemoizationListImpl.get(MemoizationList.java:99)
	at org.fxmisc.flowless.CellListManager.getCell(CellListManager.java:64)
	at org.fxmisc.flowless.CellPositioner.getSizedCell(CellPositioner.java:129)
	at org.fxmisc.flowless.VirtualFlow.getCell(VirtualFlow.java:119)
	at org.fxmisc.richtext.StyledTextArea.followCaret(StyledTextArea.java:894)
	at org.fxmisc.richtext.StyledTextArea.layoutChildren(StyledTextArea.java:1035)
	at javafx.scene.Parent.layout(Parent.java:1079)
	at javafx.scene.Parent.layout(Parent.java:1085)
	at javafx.scene.Scene.doLayoutPass(Scene.java:552)
	at javafx.scene.Scene.preferredSize(Scene.java:1646)
	at javafx.scene.Scene.impl_preferredSize(Scene.java:1720)
	at javafx.stage.Window$9.invalidated(Window.java:846)
	at javafx.beans.property.BooleanPropertyBase.markInvalid(BooleanPropertyBase.java:109)
	at javafx.beans.property.BooleanPropertyBase.set(BooleanPropertyBase.java:144)
	at javafx.stage.Window.setShowing(Window.java:922)
	at javafx.stage.Window.show(Window.java:937)
	at javafx.stage.Stage.show(Stage.java:259)
```

**Comment by TomasMikula** [comment]
If you get a `NoSuchMethodError`, that's likely because the library was compiled with one version of a dependency and you are trying to run it with another. I just republished the SNAPSHOT release to make sure it is compiled with the latest Flowless (perhaps I forgot to do this yesterday). Please try again and make sure you have the latest snapshot of RichTextFX.

**Comment by Crystark** [comment]
I removed the contents of my .m2 folder, rebuilt the project, it redownloaded flowless 0.6-SNAPSHOT and RichTextFx 1.0.0-SNAPSHOT and I'm still getting that error.

**Comment by TomasMikula** [comment]
Try also removing your `.gradle` folder.

**Comment by Crystark** [comment]
Deleted both and still having the same issue. Ttry to dig a bit but couldn't find anything yet.
I checked the jars, they indeed contain the method.

```
Caused by: java.lang.NoSuchMethodError: org.fxmisc.flowless.VirtualFlow.breadthOffsetProperty()Ljavafx/beans/property/ReadOnlyDoubleProperty;
```

How should I read that ? it tries to call `breadthOffsetProperty` on `VirtualFlow` and expects a `ReadOnlyDoubleProperty` as an answer ?

**Comment by swpalmer** [comment]
Which .gradle folder did you remove?  The one in your home directory or the one in the project directory?

Try "gradle --refresh--dependencies"

That is the correct way to read the method signature.

**Comment by TomasMikula** [comment]
And that method used to return `ReadOnlyDoubleProperty`, but now it returns `Var<Double>`. So you must be using an older version of `RichTextFX` that was compiled agains an older version of `Flowless` than what you have.

**Comment by Crystark** [comment]
This is the Jar it gets: `richtextfx-1.0.0-20160514.224554-126`

When I go into the source code, i indeed see it's returning a Var<Double>

This is the line that breaks:
```java
box.graphicOffset.bind(virtualFlow.breadthOffsetProperty());
```

Which looks ok as bind requires an `ObservableValue` which is implemented by `Val` and in turn by `Var`.

I'm clueless here. Tried at work and at home: same problem. My program is clearly reading the snapshots from the .m2 project folder as they are locked while i'm running the app. I then close the app, delete the folders, forces maven to redownload everything with the `-U` option, relaunch... same problem. I'm missing something here...

**Comment by TomasMikula** [comment]
Hmm, that is the latest published jar. Perhaps I compiled it with a not up-to-date version of Flowless. Just republished. Please, try downloading the new snapshot (`20160518.***`).

**Comment by TomasMikula** [comment]
The fix now went to 0.7-M2.
